### PR TITLE
Integrated global variables in the gval language

### PR DESCRIPTION
**Pull request by Ismail731404**

### CODE REVIEW

File: expression/gval.go [hunk 1]
```diff
@@ -3,13 +3,29 @@ package expression
...
+	listKeyValue   map[string]interface{}
+	listKeyValueMu sync.RWMutex
+}
+
+type MergedVariables struct {
+	globalVars *GlobalVariables
+	localVars  map[string]interface{}
+}
+
+const prefixGlobalVars = "global_variable_"
```
Comment (SchawnnDev) [inline expression/gval.go:24]:
```suggestion
const prefixGlobalVars = "global_"
```
maybe better ?

File: expression/gval.go [hunk 2]
```diff
@@ -3,13 +3,29 @@ package expression
...
+
 var (
+	GlobalVars *GlobalVariables = &GlobalVariables{listKeyValue: make(map[string]interface{})}
```
Comment (SchawnnDev) [inline expression/gval.go:27]:
```suggestion
	globalVars *GlobalVariables = &GlobalVariables{listKeyValue: make(map[string]interface{})}
```

this variable can be private ?

File: expression/gval.go [hunk 3]
```diff
@@ -132,3 +155,53 @@ func getEvaluable(langEval gval.Language, expression string) (gval.Evaluable, er
...
 }
+
+func (gv *GlobalVariables) Load(listKeyValue map[string]interface{}) {
+	zap.L().Info("Fetching global variables")
+
+	gv.listKeyValueMu.Lock()
+	defer gv.listKeyValueMu.Unlock()
+	for k, v := range listKeyValue {
+		gv.listKeyValue[prefixGlobalVars+k] = v
+	}
```
Comment (SchawnnDev) [inline expression/gval.go:166]:
```suggestion
	gv.listKeyValueMu.Lock()
	for k, v := range listKeyValue {
		gv.listKeyValue[prefixGlobalVars+k] = v
	}
	gv.listKeyValueMu.Unlock()
```

File: expression/gval.go [hunk 4]
```diff
@@ -132,3 +155,53 @@ func getEvaluable(langEval gval.Language, expression string) (gval.Evaluable, er
...
+
+	zap.L().Info("Global variables loaded", zap.Int("count", len(gv.listKeyValue)))
+}
+
+func (gv *GlobalVariables) Set(key string, value interface{}) {
+	gv.listKeyValueMu.Lock()
+	defer gv.listKeyValueMu.Unlock()
+
+	gv.listKeyValue[prefixGlobalVars+key] = value
```
Comment (SchawnnDev) [inline expression/gval.go:175]:
```suggestion
	gv.listKeyValueMu.Lock()
	gv.listKeyValue[prefixGlobalVars+key] = value
	gv.listKeyValueMu.Unlock()
```

File: expression/gval.go [hunk 5]
```diff
@@ -132,3 +155,53 @@ func getEvaluable(langEval gval.Language, expression string) (gval.Evaluable, er
...
+
+	zap.L().Info("Global variable set", zap.String("key", prefixGlobalVars+key), zap.Any("value", value), zap.Int("total_count", len(gv.listKeyValue)))
+}
+
+func (gv *GlobalVariables) Delete(key string) {
+	gv.listKeyValueMu.Lock()
+	defer gv.listKeyValueMu.Unlock()
+	delete(gv.listKeyValue, prefixGlobalVars+key)
```
Comment (SchawnnDev) [inline expression/gval.go:183]:
```suggestion
	gv.listKeyValueMu.Lock()
	delete(gv.listKeyValue, prefixGlobalVars+key)
	gv.listKeyValueMu.Unlock()
```

File: expression/gval.go [hunk 6]
```diff
@@ -3,13 +3,29 @@ package expression
...
 
+type GlobalVariables struct {
+	listKeyValue   map[string]interface{}
+	listKeyValueMu sync.RWMutex
+}
+
+type MergedVariables struct {
+	globalVars *GlobalVariables
+	localVars  map[string]interface{}
+}
```
Comment (SchawnnDev) [inline expression/gval.go:22]: Remove this struct, it is pretty useless

**Review by SchawnnDev — CHANGES_REQUESTED** [review]

File: expression/gval.go (hunk 1 above)
Comment (Ismail731404) [reply expression/gval.go:24]: yes.

File: expression/gval.go (hunk 2 above)
Comment (Ismail731404) [reply expression/gval.go:27]: No, unless I create a public function, the goal is to be able to update the global variable from the engine.

**Pull request merged**